vscode/src/testExplorer.ts

```diff
@@ -0,0 +1,267 @@
+  const refreshHandler = mkRefreshHandler(ctrl, context);
+  // initially populate tests
+  await refreshHandler();
+
+  ctrl.refreshHandler = refreshHandler;
+
+  const runHandler = (request: vscode.TestRunRequest) => {
+    if (!request.continuous) {
+      return startTestRun(request);
+    }
+  };
+
+  // runs an individual test run
+  // or test group (a test run where there are child tests)
+  const startTestRun = async (request: vscode.TestRunRequest) => {
+    // use the compiler worker to run the test in the interpreter
+
+    log.info("Starting test run, request was", JSON.stringify(request));
+    const worker = localGetCompilerWorker(context);
+
+    const program = await getProgramConfig();
+    if (!program) {
+      return;
+    }
+
+    for (const testCase of request.include || []) {
+      await runTestCase(ctrl, testCase, request, worker, program);
+    }
+  };
+
+  ctrl.createRunProfile(
+    "Run Tests",
+    vscode.TestRunProfileKind.Run,
+    runHandler,
+    true,
+    undefined,
+    false,
+  );
+
+  ctrl.resolveHandler = async (item) => {
+    if (!item) {
+      context.subscriptions.push(...startWatchingWorkspace(ctrl, context));
+      return;
+    }
+  };
+
+  function updateNodeForDocument(e: vscode.TextDocument) {
+    if (e.uri.scheme !== "file") {
+      return;
+    }
+
+    if (!e.uri.path.endsWith(".qs")) {
+      return;
+    }
+  }
+
+  for (const document of vscode.workspace.textDocuments) {
+    updateNodeForDocument(document);
+  }
+
+  context.subscriptions.push(
+    vscode.workspace.onDidOpenTextDocument(updateNodeForDocument),
+    vscode.workspace.onDidChangeTextDocument((e) =>
+      updateNodeForDocument(e.document),
+    ),
+  );
+}
+
+/**
+ * If there are no workspace folders, then we can't watch anything. In general, though, there is a workspace since this extension
+ * is only activated when a .qs file is opened.
+ **/
+
+function getWorkspaceTestPatterns() {
+  if (!vscode.workspace.workspaceFolders) {
+    return [];
+  }
+
+  return vscode.workspace.workspaceFolders.map((workspaceFolder) => ({
+    workspaceFolder,
+    pattern: new vscode.RelativePattern(workspaceFolder, "**/*.qs"),
+  }));
+}
+
+/**
+ * Watches *.qs files and triggers the test discovery function on update/creation/deletion, ensuring we detect new tests without
+ * the user having to manually refresh the test explorer.
+ **/
+function startWatchingWorkspace(
+  controller: vscode.TestController,
+  context: vscode.ExtensionContext,
+) {
+  return getWorkspaceTestPatterns().map(({ pattern }) => {
+    const refresher = mkRefreshHandler(controller, context, true);
+    const watcher = vscode.workspace.createFileSystemWatcher(pattern);
+    watcher.onDidCreate(async () => {
+      await refresher();
+    });
+    watcher.onDidChange(async () => {
+      await refresher();
+    });
+
+    watcher.onDidDelete(async () => {
+      await refresher();
+    });
```

Comment:
IIRC the `onDid*` methods always return objects you need to push to `context.subscriptions` so they get unsubscribed when the extension is unloaded/reloaded (best practice to avoid event handler leaks). Basically, the same pattern as the `workspace.onDid*` calls above

Reply:
I'm afraid I don't fully understand -- quick clarifying question incoming. The watcher itself is returned and added to `context.subscriptions` ([ref](https://github.com/microsoft/qsharp/pull/2059/files/7d2aadc70a74f93516ab8423c919a0cedcc4340b#diff-a3660ba9ec05210e62ad06609feb55c16cf3678ecce0b036ed0c3232a8559aaeR147)) -- are you saying that the watcher itself should _not_  get returned and instead I should return the result of the `onDid*` calls and let the watcher get dropped?

Reply:
For your reference I wrote this code based on [this example from VS Code](https://github.com/microsoft/vscode-extension-samples/blob/main/test-provider-sample/src/extension.ts#L204).

Reply:
I _was_ saying both `watcher` and these return values should get pushed the `subscriptions`... but now that I've looked at the sample code you referenced, clearly either I'm wrong or it's not important to VS Code. 

Exactly what you should do with these return values is undocumented and I've never been sure, just made my own inferences. Copying sample code is probably the safest. Feel free to resolve.

Reply:
This ended up getting removed anyway since we trigger test refreshes on `updateDocument` now